File: static/python/union_intervals.py
import sys
import collections
import random

# @include
Endpoint = collections.namedtuple('Endpoint', ('is_closed', 'val'))


class Interval:
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __lt__(self, other):
        if self.left.val != other.left.val:
            return self.left.val < other.left.val
        # Left endpoints are equal, so now see if one is closed and the other open
        # - closed intervals should appear first.
        return self.left.is_closed and not other.left.is_closed


def union_of_intervals(intervals):
    # Empty input.
    if not intervals:
        return []

    # Sort intervals according to left endpoints of intervals.
    intervals.sort()
    result = [intervals[0]]
    for i in intervals:
        if intervals and (i.left.val < result[-1].right.val or
                          (i.left.val == result[-1].right.val and
                           (i.left.is_closed or result[-1].right.is_closed))):
            if (i.right.val > result[-1].right.val or
                (i.right.val == result[-1].right.val and i.right.is_closed)):
                result[-1].right = i.right
        else:
            result.append(i)
    return result


# @exclude


def check_intervals(A):
    # Only check the intervals do not overlap with each other.
    assert all(A[i - 1].right.val < A[i].left.val or
               (A[i - 1].right.val == A[i].left.val and
                not A[i - 1].right.is_closed and not A[i].left.is_closed)
               for i in range(1, len(A)))


def main():
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 1000)
        A = []
        for _ in range(n):
            left = Endpoint(bool(random.randrange(2)), random.randrange(10000))
            right = Endpoint(
                bool(random.randrange(2)),
                random.randrange(left.val + 1, left.val + 100))
            temp = Interval(left, right)
            A.append(temp)
        ret = union_of_intervals(A)
        if ret:
            check_intervals(ret)


if __name__ == '__main__':
    main()

File: static/cpp/Union_intervals.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <random>
#include <vector>

using std::default_random_engine;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
struct Interval {
 private:
  struct Endpoint {
    bool isClosed;
    int val;
  };

 public:
  bool operator<(const Interval& i) const {
    if (left.val != i.left.val) {
      return left.val < i.left.val;
    }
    // Left endpoints are equal, so now see if one is closed and the other
    // open
    // - closed intervals should appear first.
    return left.isClosed && !i.left.isClosed;
  }

  Endpoint left, right;
};

vector<Interval> UnionOfIntervals(vector<Interval> intervals) {
  // Empty input.
  if (intervals.empty()) {
    return {};
  }

  // Sort intervals according to left endpoints of intervals.
  sort(intervals.begin(), intervals.end());
  Interval curr(intervals.front());
  vector<Interval> result;
  for (int i = 1; i < intervals.size(); ++i) {
    if (intervals[i].left.val < curr.right.val ||
        (intervals[i].left.val == curr.right.val &&
         (intervals[i].left.isClosed || curr.right.isClosed))) {
      if (intervals[i].right.val > curr.right.val ||
          (intervals[i].right.val == curr.right.val &&
           intervals[i].right.isClosed)) {
        curr.right = intervals[i].right;
      }
    } else {
      result.emplace_back(curr);
      curr = intervals[i];
    }
  }
  result.emplace_back(curr);
  return result;
}
// @exclude

void CheckIntervals(const vector<Interval>& A) {
  // Only check the intervals do not overlap with each other.
  for (size_t i = 1; i < A.size(); ++i) {
    assert(A[i - 1].right.val < A[i].left.val ||
           (A[i - 1].right.val == A[i].left.val && !A[i - 1].right.isClosed &&
            !A[i].left.isClosed));
  }
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 1000);
      n = dis(gen);
    }
    vector<Interval> A;
    for (int i = 0; i < n; ++i) {
      Interval temp;
      uniform_int_distribution<int> zero_or_one(0, 1);
      uniform_int_distribution<int> dis1(0, 9999);
      temp.left.isClosed = zero_or_one(gen), temp.left.val = dis1(gen);
      uniform_int_distribution<int> dis2(temp.left.val + 1,
                                         temp.left.val + 100);
      temp.right.isClosed = zero_or_one(gen), temp.right.val = dis2(gen);
      A.emplace_back(temp);
    }
    vector<Interval> ret = UnionOfIntervals(A);
    if (!ret.empty()) {
      CheckIntervals(ret);
    }
  }
  return 0;
}
